### PR TITLE
Add 1ad4good bidder

### PR DESCRIPTION
**Pull request by vladgurgov**
## Type of change

- [ X] New bidder adapter   
DONE: https://github.com/prebid/prebid.github.io/pull/1428

## Description of change

This is open source bidder for non profit ads that I created this weekend. Could be used to fill unsold inventory for a good causes.

More info about this project can be found here: http://1ad4good.org

I hope it will be net positive for this community as:

open source education about tech and header bidding
always filling (good for some test cases for ad ops
providing some free ad placements to non profits
Thank you!

- test parameters for validating bids
```
{
  bidder: '1ad4good',
  params: {
    placementId: 13144370
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page. -DONE

- contact email of the adapter’s maintainer vlad@aaaudi.com
- [ X] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

**Comment by vladgurgov** [comment]
@Fawke thank you for looking into it! 

It's pretty straightforward, based on appnexus bidder code(for easier adops setup) with some unused code removed. nothing special..

File: modules/1ad4goodBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+const USER_PARAMS = ['age', 'externalUid', 'segments', 'gender', 'dnt', 'language'];
+const APP_DEVICE_PARAMS = ['geo', 'device_id']; // appid is collected separately
+const DEBUG_PARAMS = ['enabled', 'dongle', 'member_id', 'debug_timeout'];
+const SOURCE = 'pbjs';
+const MAX_IMPS_PER_REQUEST = 15;
+
+export const spec = {
+  code: BIDDER_CODE,
+  aliases: ['adsforgood', 'ads4good', '1adsforgood'],
+  supportedMediaTypes: [BANNER, VIDEO, NATIVE],
```
Comment (Fawke) [inline modules/1ad4goodBidAdapter.js]: Since, you don't support Native Ads yet, I suggest you remove Native.

**Comment by vladgurgov** [comment]
Thanks for your comments @Fawke ! This was my first contribution to this project so I definitely was trying to obey all the rules.

I was developing based on appnexus adapter, so most of the code is the same.
In appnexusBidAdapter.js endpoint is 
const URL = '//ib.adnxs.com/ut/v3/prebid';

in rubicon:
export const FASTLANE_ENDPOINT = '//fastlane.rubiconproject.com/a/api/fastlane.json';
export const VIDEO_ENDPOINT = '//prebid-server.rubiconproject.com/openrtb2/auction';

in mine its:
const URL = '//hb.1ad4good.org/prebid';

Should I hardcode https:// in it? I support https, i just wanted to follow the same rules.

Tests are based on appnexusBidAdapter.js which has 75.92% coverage... I will look into adding more tests

I don't bid on video yet, but i will add video campaign on server to make it bid on video as well.

Thanks for your feedback again!

**Comment by vladgurgov** [comment]
- Unused code was removed
- Test coverages is over >80% now
- Video campaign added on bidder for testing and tested for compatibility 
- Still waiting if endpoint should be hardcoded to https:// instead of //

Thanks!

**Comment by vladgurgov** [comment]
@Fawke could you please take another look? Thank you!

**Comment by vladgurgov** [comment]
@mkendall07 @Fawke could you take look, thanks!

**Comment by bretg** [comment]
re-opening -- ping @Fawke

File: modules/1ad4goodBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+  const bid = {
+    requestId: serverBid.uuid,
+    cpm: rtbBid.cpm,
+    creativeId: rtbBid.creative_id,
+    dealId: rtbBid.deal_id,
+    currency: 'USD',
+    netRevenue: true,
+    ttl: 300,
+    adUnitCode: bidRequest.adUnitCode,
+    appnexus: {
```
Comment (Fawke) [inline modules/1ad4goodBidAdapter.js]: Is the creation of this object necessary? You might wanna get rid of this or rename it to reflect the name of your bid adapter.

File: modules/1ad4goodBidAdapter.js (hunk 1 above)
Comment (vladgurgov) [reply modules/1ad4goodBidAdapter.js]: agreed, removed NATIV

File: modules/1ad4goodBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+    Object.assign(bid, {
+      width: rtbBid.rtb.video.player_width,
+      height: rtbBid.rtb.video.player_height,
+      vastUrl: rtbBid.rtb.video.asset_url,
+      vastImpUrl: rtbBid.notify_url,
+      ttl: 3600
+    });
+
+    const videoContext = utils.deepAccess(bidRequest, 'mediaTypes.video.context');
+    if (videoContext === ADPOD) {
```
[the remark this answers is not in the document]
Comment (vladgurgov) [reply modules/1ad4goodBidAdapter.js]: Removed

File: modules/1ad4goodBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,443 @@
+import * as utils from '../src/utils';
+import { registerBidder } from '../src/adapters/bidderFactory';
+import { BANNER, NATIVE, VIDEO, ADPOD } from '../src/mediaTypes';
```
[the remark this answers is not in the document]
Comment (vladgurgov) [reply modules/1ad4goodBidAdapter.js]: Removed, thanks!

File: modules/1ad4goodBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+  supportedMediaTypes: [BANNER, VIDEO],
+
+  /**
+   * Determines whether or not the given bid request is valid.
+   *
+   * @param {object} bid The bid to validate.
+   * @return boolean True if this is a valid bid, and false otherwise.
+   */
+  isBidRequestValid: function(bid) {
+    return !!(bid.params.placementId || (bid.params.member && bid.params.invCode));
```
[the remark this answers is not in the document]
Comment (vladgurgov) [reply modules/1ad4goodBidAdapter.js]: Agreed, Removed

File: modules/1ad4goodBidAdapter.js [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+            bids.push(bid);
+          }
+        }
+      });
+    }
+
+    return bids;
+  },
+
+  getUserSyncs: function(syncOptions) {
```
[the remark this answers is not in the document]
Comment (vladgurgov) [reply modules/1ad4goodBidAdapter.js]: Removed

File: modules/1ad4goodBidAdapter.js [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+          params[convertedKey] = params[paramKey];
+          delete params[paramKey];
+        }
+      });
+    }
+
+    return params;
+  },
+
+  /**
```
[the remark this answers is not in the document]
Comment (vladgurgov) [reply modules/1ad4goodBidAdapter.js]: Nope, removed

File: modules/1ad4goodBidAdapter.js (hunk 2 above)
Comment (vladgurgov) [reply modules/1ad4goodBidAdapter.js]: renamed it to ads4good, thanks

**Comment by vladgurgov** [comment]
@Fawke , please take a look again. I removed most of unused code. 

Our initial idea with it was that we could simply use this module to test on client side easily by renaming appnexus -> 1ad4good. So that replacing one line immediately let us get fill from us and test code. And we simply wanted to make sure everything is as close to appnexus as possible. 

However, I agree with you - since we dont use most of these params it makes sense to remove it. Please take a look. Thanks!

**Comment by Fawke** [comment]
@vladgurgov 

Thanks for making the required changes. There's just one last change which I suggest you make now, you can hardcode your endpoing URL to `https` in `line:8` in `1ad4goodBidAdapter.js` file.

In the next major release of Prebid, we are having all bidder adapters convert to https. Failing to do so will result in them not being included in the next major release.

**Comment by vladgurgov** [comment]
@Fawke 
Makes sense. Just updated to https://
Please keep in mind that appnexus module still using // 

If this is accepted I can also add some sample pages and docs. I think it could be good beginners "entry-point" for Prebid and it would allow anyone to test and experiment Prebid without any accounts on Appnexus, Rubicon, their ad servers and still get 100% fill. 

Let me know if you think it makes sense of you have other ideas for contributing to this project.

Thanks!

**Pull request merged**